aws/code/workflows/MLWorkflowMultiCPU.py

```diff
@@ -0,0 +1,216 @@
+        y_test = test[label_column]
+        X_test = test.drop(label_column, axis = 1)
+
+        # persist
+        X_train = X_train.persist()
+        y_train = y_train.persist()
+
+        return X_train.astype( self.hpo_config.dataset_dtype ),\
+               X_test.astype( self.hpo_config.dataset_dtype ),\
+               y_train.astype( self.hpo_config.dataset_dtype ),\
+               y_test.astype( self.hpo_config.dataset_dtype ) 
+
+    @timer_decorator
+    def fit ( self, X_train, y_train ):       
+        """ Fit decision tree model [ architecture defined by HPO parameters ] """
+        if 'XGBoost' in self.hpo_config.model_type:
+            print('> fit xgboost model')
+            dtrain = xgboost.dask.DaskDMatrix( self.client, X_train, y_train)
+            xgboost_output = xgboost.dask.train( self.client, self.hpo_config.model_params, dtrain, 
+                                                num_boost_round = self.hpo_config.model_params['num_boost_round'] )
+            trained_model = xgboost_output['booster']
+
+        elif 'RandomForest' in self.hpo_config.model_type:
+            print('> fit randomforest model')
+            trained_model = RandomForestClassifier ( n_estimators = self.hpo_config.model_params['n_estimators'],
+                                                     max_depth = self.hpo_config.model_params['max_depth'],
+                                                     max_features = self.hpo_config.model_params['max_features'],
+                                                     n_jobs=-1 )\
+                                                     .fit( X_train, y_train.astype('int32') )                                                     
+        return trained_model 
+
+    @timer_decorator    
+    def predict ( self, trained_model, X_test, threshold = 0.5 ):
+        """ Inference with the trained model on the unseen test data """
+
+        print('> predict with trained model ')
+        if 'XGBoost' in self.hpo_config.model_type:
+            dtest = xgboost.dask.DaskDMatrix( self.client, X_test )
+            predictions = xgboost.dask.predict( self.client, trained_model, dtest)
+            predictions = (predictions > threshold ) * 1.0                    
+            
+        elif 'RandomForest' in self.hpo_config.model_type:
+            predictions = trained_model.predict( X_test )
+
+        return predictions
+
+    @timer_decorator
+    def score ( self, y_test, predictions ): 
+        """ Score predictions vs ground truth labels on test data """
+        print('> score predictions')
+        # y_test = y_test.compute()
+        score = accuracy_score ( y_test.astype( self.hpo_config.dataset_dtype ),
+                                predictions.astype( self.hpo_config.dataset_dtype ) )
+
+        print(f'\t score = {score}')
+        self.cv_fold_scores += [score]
+        return score
+
+    def save_best_model ( self, score, trained_model, filename = 'saved_model' ): 
```

Comment:
Similar to earlier comment: Do we want to determine best model on the basis of a single fold, or should we evaluate across all k?

Reply:
Yeah its an interesting question, the tricky thing is do you keep all the models in memory or do you write them all to disk? I decided to go with only saving once a new high score is reached within the cross-validation folds, though I agree its a subjective choice.

Reply:
Sorry, looking back over this, I was actually misinterpreting what we're doing with the cross-validation here. Feel free to mark this conversation as resolved, but I may touch base with you offline to make sure I'm correctly understanding the intent. That's for my own education, though; this bit looks fine.